core/arch/arm/kernel/boot.c

```diff
@@ -1252,15 +1256,55 @@ void __weak boot_init_primary_early(void)
 {
 	unsigned long pageable_part = 0;
 	unsigned long e = PADDR_INVALID;
+	struct transfer_list_entry *te = NULL;
+	struct transfer_list_header *tl = NULL;
+	struct dt_descriptor *dt = get_external_dt_desc();
 
 	if (!IS_ENABLED(CFG_WITH_ARM_TRUSTED_FW))
 		e = boot_arg_nsec_entry;
-	if (IS_ENABLED(CFG_WITH_PAGER))
-		pageable_part = boot_arg_pageable_part;
+
+	if (IS_ENABLED(CFG_TRANSFER_LIST) && transfer_list_pa) {
+		/* map the TL */
+		tl = transfer_list_map(transfer_list_pa);
+		if (!tl)
+			panic("Failed to map transfer list");
+		dt->tl = tl; /* save the mapped TL */
+
+		te = transfer_list_find(tl, TL_TAG_OPTEE_PAGABLE_PART);
+	}
+
+	if (IS_ENABLED(CFG_WITH_PAGER)) {
+		pageable_part = IS_ENABLED(CFG_TRANSFER_LIST) && te ?
+				get_le64(transfer_list_entry_data(te)) :
+				boot_arg_pageable_part;
+	}
 
 	init_primary(pageable_part, e);
 }
 
+static void boot_save_transfer_list(unsigned long transfer_list __maybe_unused,
+				    unsigned long fdt __maybe_unused)
+{
+#if defined(CFG_TRANSFER_LIST)
```

Comment:
Why is this ifdefed?

Reply:
Fixed.

Reply:
My bad, fixed.

Reply:
I need a stub function of `boot_save_transfer_list()` here for clang build without `CFG_TRANSFER_LIST`, instead I should remove `if (IS_ENABLED(CFG_TRANSFER_LIST)`.

Reply:
I would prefer having stubbed static inline functions in `transfer_list.h` when `CFG_TRANSFER_LIST` isn't defined, like for instance for `is_sp_ctx()` in `secure_partition.h`.

Reply:
Done.